Assets/MixedRealityToolkit.Providers/XRSDK/XRSDKDeviceManager.cs

```diff
@@ -0,0 +1,244 @@
+
+using Microsoft.MixedReality.Toolkit.Input;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace Microsoft.MixedReality.Toolkit.XRSDK.Input
+{
+    /// <summary>
+    /// Manages XR SDK devices.
+    /// </summary>
+    [MixedRealityDataProvider(
+        typeof(IMixedRealityInputSystem),
+        SupportedPlatforms.WindowsStandalone | SupportedPlatforms.MacStandalone | SupportedPlatforms.LinuxStandalone | SupportedPlatforms.WindowsUniversal,
+        "XRSDK Device Manager")]
+    public class XRSDKDeviceManager : BaseInputDeviceManager, IMixedRealityCapabilityCheck
+    {
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="inputSystem">The <see cref="Microsoft.MixedReality.Toolkit.Input.IMixedRealityInputSystem"/> instance that receives data from this provider.</param>
+        /// <param name="name">Friendly name of the service.</param>
+        /// <param name="priority">Service priority. Used to determine order of instantiation.</param>
+        /// <param name="profile">The service's configuration profile.</param>
+        public XRSDKDeviceManager(
+            IMixedRealityInputSystem inputSystem,
+            string name = null,
+            uint priority = DefaultPriority,
+            BaseMixedRealityProfile profile = null) : base(inputSystem, name, priority, profile) { }
+
+        /// <inheritdoc />
+        public virtual bool CheckCapability(MixedRealityCapability capability)
+        {
+            // The XR SDK platform supports motion controllers.
+            return (capability == MixedRealityCapability.MotionController);
+        }
+
+        protected static readonly Dictionary<string, GenericXRSDKController> ActiveControllers = new Dictionary<string, GenericXRSDKController>();
+
+        private InputDevice leftInputDevice;
+        private InputDevice rightInputDevice;
+        private bool wasLeftInputDeviceValid = false;
+        private bool wasRightInputDeviceValid = false;
+
+        /// <inheritdoc/>
+        public override void Update()
+        {
+            base.Update();
+
+            if (XRSDKSubsystemHelpers.InputSubsystem == null || !XRSDKSubsystemHelpers.InputSubsystem.running)
+            {
+                return;
+            }
+
+            if (!leftInputDevice.isValid)
+            {
+                if (wasLeftInputDeviceValid)
+                {
+                    GenericXRSDKController controller = GetOrAddController(leftInputDevice);
+
+                    if (controller != null)
+                    {
+                        CoreServices.InputSystem?.RaiseSourceLost(controller.InputSource, controller);
+                    }
+
+                    RemoveController(leftInputDevice);
+                    wasLeftInputDeviceValid = false;
+                }
+
+                leftInputDevice = InputDevices.GetDeviceAtXRNode(XRNode.LeftHand);
+
```

Comment:
Reading this is a little funky, because of this:

1) Above there's a check to see if the input device isn't valid
2) Then it removes it if it's not valid.
3) Now I re-get the left input device
4) It seems weird that it would suddenly be valid

But I think this makes sense if the left hand in this case corresponds to some other "left hand" input, which existed, showed up, was destroyed, and then a new left hand was created. (i.e. this is effectively a new InputDevice, and the old one will forever be !isValid because it was ded).

Is this the case? If so it might be helpful just to have a comment on this (i.e. because the old one makes permanent state transitions to invalid, we have to within the same scope then look for a new one).

Reply:
Yeah, basically:

1. Check if the previously seen (cached) input device is still valid
1. Remove it if it's not valid (lost source, hand out of view, etc)
1. Then ask Unity if there's a new valid source for the specified hand

Whenever a hand is lost / detected, it's considered a new input device, as far as I've seen.

>because the old one makes permanent state transitions to invalid, we have to within the same scope then look for a new one

Yep that exactly